OemPkg/SecureBootRecovery/SecureBootRecovery.c

```diff
@@ -0,0 +1,136 @@
+/** @file
+  This application will attempt to append the 'Windows UEFI CA 2023' and then reboot the system.
+  On success, this application will allow the system to boot into 2023 signed Windows
+
+  Copyright (C) Microsoft Corporation
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+**/
+
+#include <Uefi.h>
+#include <Library/UefiApplicationEntryPoint.h>
+#include <Guid/ImageAuthentication.h>
+
+#include "RecoveryPayload.h"
+
+// 10 seconds in microseconds
+#define STALL_10_SECONDS  10000000
+
+/**
+  The user Entry Point for Application. The user code starts with this function
+  as the real entry point for the application.
+
+  @param[in] ImageHandle        The firmware allocated handle for the EFI image.
+  @param[in] SystemTable        A pointer to the EFI System Table.
+
+  @retval EFI_SUCCESS           The entry point is executed successfully.
+  @retval EFI_INVALID_PARAMETER SystemTable provided was not valid.
+  @retval other                 Some error occurs when executing this entry point.
+
+**/
+EFI_STATUS
+EFIAPI
+UefiMain (
+  IN EFI_HANDLE        ImageHandle,
+  IN EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  EFI_STATUS  Status;
+  UINT32      Attributes;
+
+  Attributes = EFI_VARIABLE_NON_VOLATILE |
+               EFI_VARIABLE_BOOTSERVICE_ACCESS |
+               EFI_VARIABLE_RUNTIME_ACCESS |
+               EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS |
+               EFI_VARIABLE_APPEND_WRITE;
+
+  //
+  // Start checking that the system is in a state we can safely use
+  //
+  if (SystemTable == NULL) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if ((SystemTable->ConOut == NULL) || (SystemTable->ConOut->OutputString == NULL) || (SystemTable->ConOut->ClearScreen == NULL)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  if ((SystemTable->BootServices == NULL) || (SystemTable->BootServices->Stall == NULL)) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  //
+  // After this point, we should be able to print and stall but nothing else has been verified
+  //
+  if ((SystemTable->RuntimeServices == NULL) || (SystemTable->RuntimeServices->ResetSystem == NULL)) {
+    Status = EFI_INVALID_PARAMETER;
+    goto Exit;
+  }
+
+  //
+  // Start informing the user of what is happening
+  //
+  Status = SystemTable->ConOut->ClearScreen (SystemTable->ConOut);
+  if (EFI_ERROR (Status)) {
+    goto Exit;
+  }
+
+  Status = SystemTable->ConOut->OutputString (SystemTable->ConOut, L"\r\nAttempting to update the system's secureboot certificates\r\n");
+  if (EFI_ERROR (Status)) {
+    goto Exit;
+  }
+
+  SystemTable->ConOut->OutputString (SystemTable->ConOut, L"Learn more about this tool at https://aka.ms/securebootrecovery\r\n");
+
+  //
+  // Perform the append operation
+  //
+  Status = SystemTable->RuntimeServices->SetVariable (
+                                           L"db",
+                                           &gEfiImageSecurityDatabaseGuid,
+                                           Attributes,

```

Comment:
Are there any concerns that some systems could store the `db` with different attributes than we are setting here? I know they shouldn't, but at the same time, I wonder if anyone is...in which case would it be better to do a `GetVariable` on `db` first to ensure the attributes are what we expect and if not either using the retrieved attributes or deleting the variable and overwriting it with the data retrieved from `GetVariable`? Just trying to think about making this work on any given system, the worst case ones, etc.

Reply:
If they are then the SetVariable(..) would fail, the attributes are part of the signature digest on the authenticated variable payload. If this happens then there are very few things we can do to resolve it, fewer we want to do pre ExitBootServices(..)

Reply:
Right, that was my point of SetVariable failing and if we should try to recover with it. I didn't know that the attributes are part of the signature digest, so seems like we think this really should not happen. Just wanted to check, since this is critical to booting, whether we needed to get the existing db, delete the variable, write it again with the correct attributes. Sounds like no.

Reply:
Side note, but people should be using variable policy to constrain the variable to an expected set of attributes.